File: dotnet/Allmark/Inline/AutolinkRule.cs
namespace Allmark.Inline;

using System.Text.RegularExpressions;
using Allmark.Types;

public static class AutolinkRule
{
    private static readonly Regex SpaceRegex = new(@"\s");
    private static readonly Regex LinkRegex = new(@"^<(\s*[a-z][a-z0-9+.-]{1,31}:[^<>]*)>", RegexOptions.IgnoreCase);
    private static readonly Regex EmailRegex = new(
        @"^<(\s*[a-z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*\s*)>",
        RegexOptions.IgnoreCase);

    public static InlineRule Create()
    {
        return new InlineRule
        {
            Name = "autolink",
            Test = TestAutolink,
        };
    }

    private static bool TestAutolink(InlineParserState state, MarkdownNode parent)
    {
        // Don't try to extract HTML for HTML blocks
        if (parent.Type == "html_block")
        {
            return false;
        }

        var ch = Utils.GetChar(state.Src, state.I);
        if (ch == '<' && !Utils.IsEscaped(state.Src, state.I))
        {
            var tail = state.Src.Substring(state.I);

            var linkMatch = LinkRegex.Match(tail);
            if (linkMatch.Success)
            {
                var url = Utils.EscapeHtml(linkMatch.Groups[1].Value);

                if (SpaceRegex.IsMatch(url))
                {
                    var content = Utils.EscapeHtml(linkMatch.Groups[0].Value);
                    var text = Utils.NewText(state.ParentIndex + state.I, state.Line, content, state.Indent);
                    text.Length = linkMatch.Groups[0].Length;
                    parent.Children!.Add(text);
                    state.I += linkMatch.Groups[0].Length;

                    return true;
                }

                var escapedUrl = url.Replace("\\", "\\\\");
                var decodedUrl = Utils.DecodeEntities(url);
                var encodedUrl = Utils.EscapeUriString(decodedUrl);

                var linkText = Utils.NewText(state.ParentIndex + state.I, state.Line, escapedUrl, state.Indent);

                var link = Utils.NewInline("link", state.ParentIndex + state.I, state.Line, "", state.Indent);
                link.Info = encodedUrl;
                link.Length = linkMatch.Groups[0].Length;
                link.Children = [linkText];
                parent.Children!.Add(link);
                state.I += linkMatch.Groups[0].Length;

                return true;
            }

            var emailMatch = EmailRegex.Match(tail);
            if (emailMatch.Success)
            {
                var url = Utils.EscapeHtml(emailMatch.Groups[1].Value);

                if (SpaceRegex.IsMatch(url))
                {
                    var content = Utils.EscapeHtml(emailMatch.Groups[0].Value);
                    var text = Utils.NewText(state.ParentIndex + state.I, state.Line, content, state.Indent);
                    text.Length = emailMatch.Groups[0].Length;
                    parent.Children!.Add(text);
                    state.I += emailMatch.Groups[0].Length;

                    return true;
                }

                var decodedUrl = Utils.DecodeEntities(url);
                var encodedUrl = Utils.EscapeUriString(decodedUrl);

                var linkText = Utils.NewText(state.ParentIndex + state.I, state.Line, url, state.Indent);

                var link = Utils.NewInline("link", state.ParentIndex + state.I, state.Line, "", state.Indent);
                link.Info = $"mailto:{encodedUrl}";
                link.Length = emailMatch.Groups[0].Length;
                link.Children = [linkText];
                link.Children = [linkText];
                parent.Children!.Add(link);
                state.I += emailMatch.Groups[0].Length;

                return true;
            }
        }

        return false;
    }
}

File: web/src/inline/autolinkRule.ts
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import { ANGLE_LEFT_CODE } from "../utils/charCodes";
import decodeEntities from "../utils/decodeEntities";
import escapeHtml from "../utils/escapeHtml";
import isEscaped from "../utils/isEscaped";
import newInline from "../utils/newInline";
import newText from "../utils/newText";

const rule: InlineRule = {
	name: "autolink",
	test: testAutolink,
};
export default rule;

// An HTML tag consists of an open tag, a closing tag, an HTML comment, a
// processing instruction, a declaration, or a CDATA section.
const SPACE_REGEX = /\s/;
const LINK_REGEX = /^<(\s*[a-z][a-z0-9+.-]{1,31}:[^<>]*)>/i;
const EMAIL_REGEX =
	/^<(\s*[a-z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*\s*)>/i;

function testAutolink(state: InlineParserState, parent: MarkdownNode): boolean {
	// Don't try to extract HTML for HTML blocks
	if (parent.type === "html_block") {
		return false;
	}

	if (state.src.charCodeAt(state.i) === ANGLE_LEFT_CODE && !isEscaped(state.src, state.i)) {
		let tail = state.src.substring(state.i);

		let linkMatch = tail.match(LINK_REGEX);
		if (linkMatch !== null) {
			let url = escapeHtml(linkMatch[1]);

			if (SPACE_REGEX.test(url)) {
				let content = escapeHtml(linkMatch[0]);
				let text = newText(state.parentIndex + state.i, state.line, content, state.indent);
				text.length = linkMatch[0].length;
				parent.children!.push(text);
				state.i += linkMatch[0].length;

				return true;
			}

			let text = newText(
				state.parentIndex + state.i,
				state.line,
				url.replaceAll("\\", "\\\\"),
				state.indent,
			);
			let link = newInline("link", state.parentIndex + state.i, state.line, "", state.indent);
			link.children = [text];

			url = decodeEntities(url);
			url = encodeURI(decodeURI(url));

			link.info = url;
			link.length = linkMatch[0].length;
			parent.children!.push(link);

			state.i += linkMatch[0].length;

			return true;
		}

		let emailMatch = tail.match(EMAIL_REGEX);
		if (emailMatch !== null) {
			let url = escapeHtml(emailMatch[1]);

			if (SPACE_REGEX.test(url)) {
				let content = escapeHtml(emailMatch[0]);
				let text = newText(state.parentIndex + state.i, state.line, content, state.indent);
				text.length = emailMatch[0].length;
				parent.children!.push(text);
				state.i += emailMatch[0].length;

				return true;
			}

			let text = newText(
				state.parentIndex + state.i,
				state.line,
				url.replaceAll("\\", "\\\\"),
				state.indent,
			);
			let link = newInline("link", state.parentIndex + state.i, state.line, "", state.indent);
			link.children = [text];

			url = `mailto:${encodeURI(url)}`;

			link.info = url;
			link.length = emailMatch[0].length;
			parent.children!.push(link);

			state.i += emailMatch[0].length;

			return true;
		}
	}

	return false;
}
